### ISSUE TITLE
Fire "beforeunload" confirm box only if document has changes

### ISSUE THREAD

**Issue by sdumetz** [Smithsonian/dpo-voyager#205]
This box is annoying. The save process is complicated enough (save scene, capture pose, save article), if users can't rely on this to _know_ if they forgot to save something they will just ignore the dialog because it's wrong most of the time anyway.

Serializing the `CVDocument` just after `openDocument()` and serializing again to compare the string value in `CVStoryApplication.beforeunload()` might do the trick but I'd like a solution that allows a better UI (eg: highlighting the **save** button as soon as the doc is changed). The tricky bit is that we have no representation of the document's data once it's been loaded into the graph, so a "changed" switch seems hard to implement.

I'd like to work on this issue, if anyone has an idea worth trying.

**Comment by gjcope** [comment]
I agree the box would be better served only when the data is dirty, but still better there than not. Several components already have literal or effective 'changed' output events, so a state manager could subscribe to those events to flip the global changed state and clear on save. The hard part would be adding the missing events and making sure coverage is 100%.

**Comment by sdumetz** [comment]
Reviving this thread. I see two possible paths forward, detailed below.

## Check if serialized state has changed

I made a prototype (https://github.com/Holusion/dpo-voyager/tree/modified_warning) that performs a comparison between the original document and the serialized state.

It's temporarily bound to log when clicking on the **download** button to be easier to test but its intended use is within the `beforeunload` handler.

It kind of works. Putting aside edge cases, the main drawback is that  we can't know if the document is dirty before the method is triggered (eg: to highlight the save button, showing there is _something_ to save). We can't really continuously serialize and compare the application state _just in case_.

## Emit "dirty" events

### From properties

Deserialized data is overwhelmingly (Exception: `CVMeta`'s collections) put into `Component.ins.[...]` as instances of the `Property` class. Ideally that's where we would surface changes.

But properties have no way to "know" whether or not (and how) they are serialized. They could still be made to contain a `dirty` flag that would behave like a persistent version of their `changed` flag, but the implementation logic would have to be elsewhere

Prototype implementation of a `dirty` flag on properties: https://github.com/Holusion/dpo-voyager/commit/e03415d750f78e3a051fb1938333d1c69b73bb30. The hard part would be to propagate this through each parent components.

### From individual CV* components 

Would be better from a UX perspective as it allow us to be proactive (highlighting the save button when state is dirty).

Here is a list of what needs components would need to be checked for "dirtiness" : 
- **NVScene**
  - CVScene
  - CVSetup
    - CVInterface;
    - CVReader;
    - CVViewer;
    - CVOrbitNavigation;
    - CVDerivativesController;
    - CVBackground;
    - CVFloor;
    - CVGrid;
    - CVTape;
    - CVSlicer;
    - CVTours;
    - CVSnapshots;
    - CVEnvironment;
    - CVLanguageManager;
    - CVAudioManager;
    - CVActionManager;
  - CVMeta
  - **NVNode**
    - CVMeta
     - UnorderedCollection<any> collection
    - UnorderedCollection<any> process
    - UnorderedCollection<IImage> images
    - OrderedCollection<Article> articles
    - INote[] notes
    - UnorderedCollection<IAudioClip> audio
    - UnorderedCollection<IAction> actions
    - CVModel2
    - CVCamera
    - CLight (and all `C*Light` and `CV*Light`  descendants)

That's quite a long list (still probably missing a few). On the other hand most of those should be relatively straightforward.

One way that can be done is adding a `firstUpdated` flag to the `Component` class (similar to `CustomElement.firstConnected()`) and edit each individual `CV*` class' `update()` method with something like:

```javascript
if(this.firstUpdated && (ins.foo.changed || [...]){
    if(!this.dirty) this.emit("dirty")
    this.dirty = true;
}
```

**Comment by gjcope** [comment]
> But properties have no way to "know" whether or not (and how) they are serialized. They could still be made to contain a dirty flag that would behave like a persistent version of their changed flag, but the implementation logic would have to be elsewhere

What about adding a new 'serializeProperties' list (or something of the like) similar to the existing 'settingProperties' and 'snapshotProperties' to components. This should allow the dirty check to be centralized and result in a dirty event that can be watched for. I realize, like the "From individual CV* components" proposal that this also requires adding something to each of the identified classes but I wonder if it might be more obvious for future sustainability.

**Comment by sdumetz** [comment]
> What about adding a new 'serializeProperties' list (or something of the like) similar to the existing 'settingProperties' and 'snapshotProperties' to components. This should allow the dirty check to be centralized and result in a dirty event that can be watched for. I realize, like the "From individual CV* components" proposal that this also requires adding something to each of the identified classes but I wonder if it might be more obvious for future sustainability.

That would probably be the way to go if trying to list the relevant properties in each individual component. Functionally the same as writing a big `if( ins.something.dirty [...])`, but cleaner, yes.

Though it's a slippery slope as it requires never forgetting to add a property to this list when a new one is created or a behavior is changed. There would be no systematic way to check if we're missing a few obscure edge cases.

The big advantage of the Serialization approach is that once it works for all "save states" (Story, QC, "saved setup"), most remaining bugs should be false positives (eg: in the current master, camera coordinates are serialized even though `navigation` properties always takes precedence). So worst case is we tell a user he should save when he has no reason to. It's far better than the other way around.

In the end it would come down to how much code complexity and risk of not warning the user properly we are willing to have to be able to get this `document-dirty` event.

I'd love to have it. I even have (kind of) the infrastructure to implement an auto-save feature from it, which a lot of users would love to have.

But I'm not sure the tradeoffs are good. I've spent way more time exploring the "emit dirty event" route than "serialize the document". Yet I've got nothing to show for the former, and a working prototype for the later.

My concern is not merely the amount of work to list all the relevant properties in all the relevant components. I think the amount of code required to map this list to a system emitting events is huge. Then we have no tools to evaluate coverage: it's all manual testing (initially and in the future). And once everything is said and done, it would be one-way: we'd need to store all properties' initial values to detect if a document ceases to be dirty. Wouldn't we want that, too? Because the Serialization route provides this out of the box.

**Comment by gjcope** [comment]
> Though it's a slippery slope as it requires never forgetting to add a property to this list when a new one is created or a behavior is changed. There would be no systematic way to check if we're missing a few obscure edge cases.

True, but not much different from the other settings lists mentioned. If it's not added, the functionality doesn't work as expected. Not ideal from a development standpoint but makes for a better user experience. I was thinking about using the base fromData() to set up the list, but unfortunately not all of the components are a one-to-one mapping.

> And once everything is said and done, it would be one-way: we'd need to store all properties' initial values to detect if a document ceases to be dirty. Wouldn't we want that, too? 

I don't think this is true. With typical event-driven systems that track dirty states you would also have a reset event. So every time the document is saved, the manager (in this case probably CVStoryApplication) resets its state and emits a reset event that the other components subscribe to.

That all being said, I'm not opposed to the serialization option. I just think an actively tracked dirty state would be a value add rather than just addressing the save warning annoyance. But if your implementation is performant with relatively large scene files, let's give it a go. It definitely requires less maintenance and addresses a common complaint so deems like a decent ROI.

**Comment by sdumetz** [comment]
> I don't think this is true. With typical event-driven systems that track dirty states you would also have a reset event. So every time the document is saved, the manager (in this case probably CVStoryApplication) resets its state and emits a reset event that the other components subscribe to.

I was thinking about when the user reverts his changes manually (typ. a switch toggled twice). Resetting would be easy.

Anyways, thanks for your inputs as always. If I can find a way to make the implementation itself elegant enough, I'd be OK with one hour or two of copy-pasting property names. I'll see what I can come up with.
